ENH: Add annotations for `np.core.arrayprint`

BvB93 opened the pull request:
Per the title: this PR adds annotations for all array-printing-related functions in `np.core.arrayprint`.

@charris continuing from https://github.com/numpy/numpy/pull/7859#issuecomment-747588637: 
The signature of `np.set_printoptions` can quite easily be annotated as it turns out,
so annotations can definitely help here with aforementioned issue.

Commit:
DOC: Clarify that `np.array2string` can only take an `ndarray`

Commit:
ENH: Added annotations for `np.core.arrayprint`

Commit:
TST: Added typing tests for `np.core.arrayprint`

BvB93 commented on numpy/core/arrayprint.pyi:
```diff
@@ -0,0 +1,148 @@
+    void: Callable[[void], str]
+    numpystr: Callable[[_CharLike], str]
+    object: Callable[[object], str]
+    str: Callable[[Any], str]  # Unused but still present?
```
The docstring of `np.setprintoptions` ([doc](https://github.com/numpy/numpy/blob/bac54ecb0b815eea2e8116d21aa96e1387793468/numpy/core/arrayprint.py#L132-L157)) mentions an optional `str` key, though for the past 8 years it has actually been non-functional (see https://github.com/numpy/numpy/pull/459).

Can we just remove any mention of it, both here and in the docstring?

rgommers replied:
Sounds fine to me.

BvB93 replied:
Done as of https://github.com/numpy/numpy/pull/18032/commits/82e5747dd40cbbde9c728452f21fa4716492c81d.

Just to stress: the above-mentioned push does not deprecate anything, 
it merely removes a few lines of unused code (and updates the docstring).

eric-wieser commented on numpy/core/arrayprint.pyi:
```diff
@@ -0,0 +1,148 @@
+    formatter: Optional[_FormatDict] = ...,
+    sign: Optional[Literal["-", "+", " "]] = ...,
+    floatmode: Optional[
+        Literal["fixed", "unique", "maxprec", "maxprec_equal"]
```
This might be worth abstracting to a `_FloatMode =  Literal["fixed", "unique", "maxprec", "maxprec_equal"]`

BvB93 replied:
Not a bad idea, done as of https://github.com/numpy/numpy/pull/18032/commits/05fa6e4bc219fb4db0808ef5cf9aed96c8ec0844.

Commit:
MAINT: Remove any mention of `formatter["str"]`

It's been non-functional for the past 8 years (xref https://github.com/numpy/numpy/pull/459)

Commit:
MAINT: Added a type-alias for the `floatmode` annotations

Co-Authored-By: Eric Wieser <425260+eric-wieser@users.noreply.github.com>

charris commented:
Thanks Bas.
